Push notifications

coryzoon opened the pull request:
The feature is for push notifications. At a high-level the code updates the settings screen, adds a new notification landing page screen, adds Firebase, and lays the foundation for push notifications and subscribing/unsubscribing to topics in Firebase.

- Added and configured google firebase (the changes you'll see in the */ios/* and */android/* directories).
- Additional code that removes the AD_ID permissions that will be added by the firebase library.
- Added the following libraries:
            firebase (analytics, app, messaging) - for push notifications and analytics.
            reduxjs/toolkit - state management for settings and storing past notifications.
       react-native-linear-gradient - for gradient buttons in the designs.
       react-native-dotenv - env file to store the sendgrid api token for email the testing tokens.
- Updated classifiers to toggle on the push notification settings after a user classifies on a new project.
- Sends an email to android@zooniverse.org when a tester logs in and has push notifications enabled.
- Updated Navbar.js to allow for navigation headers that use a zoon icon in the upper right.
- NotificationLandingPageScreen.js is a new screen that user is navigated to when they click on a push notification. It displays an archive of notifications. Added link to this screen to the sidebar. Added some additional components to be used in this screen.
- Updated Settings.js screen to work with the new push notifications code. 
- Added redux slices for the notifications and the notification settings.
- PushNotifications.js is a class that handles all of the push notifications work. It populates the settings, sets the defaults, handles when a user classifies on a project, emails the testing push token, handles asking the user for permission to enable push notificatinos, handles when settings are toggled (update state, sub/unsub the topic to firebase), handles incoming push notifications, and saves them to storage.
- Added notifications/helpers.js file to handle converting the timestamp to a human readable format.
- testers.js file lists the testers logins that should have the notification push token emailed to them upon logging in.

Update: All of the hound[bot] warnings that say "Parsing error: 'import' and 'export' may appear only with 'sourceType: module'" should be fixed with my latest push.

Designs: Can be found at https://www.figma.com/file/QC931orjhCdGoQogVg7iDZ/Mobile-App?type=design&node-id=134-764&mode=design under "Notifications Landing Page"

Testing: Please reference https://docs.google.com/document/d/1kp6wJPu0R1nq8LX49l_BzgQrliqcOoj_1wNxKy04uY4/edit#heading=h.31ze87hybrn on how to get a testing push token and https://docs.google.com/document/d/1sFok_Rp1mMrYGjI2aFxNXBGFo5erln-QGc3Ptya-aok/edit#heading=h.x2exepry2h4p on how to send a test push notification. If you want to login as a tester and message me I can grab your testing token from email and send you some test push messages so you don't have to.

mcbouslog commented:
Is there mock data I could plug into somewhere to see the components (most notably the components in `components/notifications`) as intended?

Are there (presumably Figma) designs I can reference?

mcbouslog reviewed (changes requested):
This is looking good, very exciting to see notifications related work coming along so well!

Might just be me, but I had to run `pod repo update` then `pod install` to get iOS working.

I only did a cursory review of `/android` and `/ios` changes, as I'm generally not familiar, but feel issues will come up in testing and deployment.

PushNotifications looks good, and no individual method is particularly complex, but it's a large class, I'm not sure if it could be broken up. Not necessary, but for consideration.

The `getAllUserClassifications` function could cause an issue (as noted in src/api/index.js) with users with a lot of classifications. I think there are other approaches that might be preferable, like how we get the projects classified for the [user profile ribbon](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/pages/home-for-user/index.jsx#L114) on the logged in user homepage or using the new [stats service ERAS](https://github.com/zooniverse/eras/wiki/API-Callout-Examples#example-requests-2) (notably with `?project_contributions=true`).

I did not visually review new components in `components/notifications` or other components reflecting actual notifications yet, not sure what the best way to do that is (i.e. I add mock data locally, hardcode my user as a tester similar to how Cory has, using a zootester account, or otherwise).

mcbouslog commented:
Probably not related to this PR, but `npm test` isn't working for me, might not have been working for awhile (like years?). I don't think any testing related work needs to be done with this PR, but maybe it'd be a good idea to open a related issue (to get the basic test script working) and then consider the most minimal of tests for the new notifications functionality.

coryzoon commented:
> Is there mock data I could plug into somewhere to see the components (most notably the components in `components/notifications`) as intended?
> 
> Are there (presumably Figma) designs I can reference?



> Is there mock data I could plug into somewhere to see the components (most notably the components in `components/notifications`) as intended?
> 
> Are there (presumably Figma) designs I can reference?

There is no mock data but I've updated the description with testing instructions. Please reach out to me if you'd like for me to send the test push messages. I've also added the figma link to the description.

coryzoon commented on src/notifications/testers.js:
```diff
@@ -0,0 +1,17 @@
+ * push notification token to the listed email. This is used
+ * for testing push notifications before they are sent to everyone.
+ */
+export const pushTesters = [
```
Thanks for the feedback, I broke down the large class so it should be a little smaller now. As far as the user classifications those should be running asynchronously and should not block the user, however, if you have a user that has a lot of classifications that I could test with, it would help me confirm. Also I had a discussion with @lcjohnso  about the classifications and I believe it was agreed to not use the ERAS endpoint at this time (Cliff, please correct me if I misunderstood this). I get the classifications anytime the project list on the home screen is refreshed, so I want to make sure I'm avoiding any potential performance issues.

You can use a zootester login and/or I can add you to the tester list if you want to send me your username and the email you would want the test token sent to. (We can jump on a call if needed and I can walk your through everything and make sure you have what you need to test/PR).

coryzoon commented on src/components/ProjectDisciplines.js:
```diff
@@ -138,11 +139,9 @@ function ProjectDisciplines({ ...props }) {
-        );
-        props.settingsActions.addUnusedProjectsToNotifications(
-          notificationProjects
+          projectList.filter((project) => !project.isPreview).filter( project => !project.isPreview && project.launch_approved )
```
Good catch, fixed.

coryzoon commented on src/components/settings/Settings.js:
```diff
@@ -1,189 +1,138 @@
+        <SettingHeader text="Notification Settings" />
+      </View>
+      <FontedText style={styles.disclosureText}>
+        The Zooniverse app can occassionally send you updates about new projects
```
Good catch, looks like it had been this way for years, fixed.

coryzoon commented on src/api/index.js:
```diff
@@ -0,0 +1,30 @@
+
+      break;
+    } catch (err) {
+      throw new Error('Error getting all user classifications', err.message);
```
Updated.

coryzoon commented on src/notifications/PushNotifications.js:
```diff
@@ -0,0 +1,370 @@
+        'true'
+      );
+    } catch (e) {
+      throw new Error(
```
Updated.

coryzoon commented on src/containers/app.js:
```diff
@@ -38,6 +38,7 @@ import PageKeys from '../constants/PageKeys'
 import AsyncStorage from "@react-native-async-storage/async-storage";
 import * as Sentry from '@sentry/react-native';
+import { PushNotifications } from '../notifications/PushNotifications';
 
```
Updated with imports removed.

coryzoon commented on src/components/NotificationLandingPageScreen.js:
```diff
@@ -0,0 +1,170 @@
+  const checkedForProjects = useRef(false); // Used to ensure the setTimeout is only run once.
+
+  const filteredProjects = extractSwipeEnabledProjects(
+    projectList.filter((project) => !project.isPreview).filter( project => !project.isPreview && project.launch_approved )
```
Updated.

coryzoon commented:
> Probably not related to this PR, but `npm test` isn't working for me, might not have been working for awhile (like years?). I don't think any testing related work needs to be done with this PR, but maybe it'd be a good idea to open a related issue (to get the basic test script working) and then consider the most minimal of tests for the new notifications functionality.

It's likely been like this for a while. I agree that it would be good to open a related issue to get the existing tests passing and then add new tests if possible.

coryzoon commented:
@mcbouslog I believe I have replied to each of your comments and made any needed updates. Please reach out if you would like to go over this in more detail or if you need any help with testing.

lcjohnso commented on src/api/index.js:
```diff
@@ -0,0 +1,30 @@
+import apiClient from 'panoptes-client/lib/api-client';
+
+export const getAllUserClassifications = async (userId) => {
```
Copying @coryzoon comment relevant to this discussion into this thread:
> As far as the user classifications those should be running asynchronously and should not block the user, however, if you have a user that has a lot of classifications that I could test with, it would help me confirm. Also I had a discussion with @lcjohnso about the classifications and I believe it was agreed to not use the ERAS endpoint at this time (Cliff, please correct me if I misunderstood this). I get the classifications anytime the project list on the home screen is refreshed, so I want to make sure I'm avoiding any potential performance issues.

Clarifying past discussions with Cory about stats: stats came up in regards to two issues related to the user ribbon (https://github.com/zooniverse/mobile/issues/265; primary issue for documentation purposes) and classification count updates (https://github.com/zooniverse/mobile/issues/216). My statement at the time was that we should prioritize other work over stat reporting improvements and a migration to ERAS.  While I was aware of Cory's use of the API to query classification counts and identify new projects where the user has participated, I didn't realize / comprehend what endpoint Cory was querying and the potential issue regarding users with large classification counts.

I agree with Mark -- the current implementation that uses the classification endpoint is not a workable solution. One alternative would be to mimic the method used by the user ribbon, querying the user project preferences (where one iterates over projects rather than individual classification), however even this has been shown to run slow for superusers.  My recommendation would be to use the new ERAS service instead of the Panoptes API's classification endpoint -- specifically the authenticated user classification endpoint (see https://github.com/zooniverse/eras/wiki/API-Callout-Examples#example-requests-2)

Example call from those docs:

    ## query to get total_count and project_contributions (and counts per project) of all classifications of user with id 1
    GET https://eras-staging.zooniverse.org/classifications/users/1?project_contributions=true

coryzoon replied:
I pushed a change that uses the ERAS endpoint above. I'm going to continue working on this to see if i can use the panoptes-client to achieve this instead of an HTTP request to the endpoint. @mcbouslog I looked at the frontend code base for a similar implementation but couldn't find anything. If you are aware of any snippets that might help me with this, please let me know.

lcjohnso replied:
@coryzoon -- Check out the `eras-client` that is available within the panoptes-javascript-client: https://github.com/zooniverse/panoptes-javascript-client/blob/main/lib/eras-client.js

That library is used in the Panoptes-Front-End for project stats pages ([link to code](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/pages/project/stats/stats.jsx)), so might be helpful to you.

coryzoon replied:
@lcjohnso Looking at https://github.com/zooniverse/eras/wiki/API-Callout-Examples#example-requests-2 it seems like the example call you pasted above is exactly what I need as it gives me all of the project ids that the user has classified on. I'm not seeing that panoptes-client supports this endpoint. The example above looks like it's trying to get /classifications?workflow_id=x&period=x whereas I'm trying to get /classifications/user/x?project_contributions=true. In eras-client.js I don't see where it would pass the project_contributions=true param to the http request. Maybe @mcbouslog can confirm if this should be possible or not. I do currently have it working without using the panoptes-client so if it's not supported yet we could proceed with what I have and then update it later once it's available in panoptes-client.

lcjohnso replied:
There's no need to use `panoptes-client`/`eras-client`, but wanted you to be aware of that piece of the client if only just as a reference for your custom code.

@mcbouslog might want to comment on whether he will be wanting to use `eras-client` to query `/classifications/users/X` with `project_contributions` query param -- perhaps it is worth it to add that funcationality now to the `panoptes-client` for use in mobile and later for FEM user stats pages.  But from my perspective, it is also OK if Cory writes one-off code for use in mobile that doesn't contribute to `eras-client` -- we don't necessarily need to make all client-like code reusable.

mcbouslog replied:
`panoptes-client/eras-client` doesn't support the noted `project_contributions` param or credentialed requests yet (`/users/:id` or `/groups/:id`), which I think are needed for this `mobile` purpose. That functionality will be added eventually, but I agree these one-off changes are fine for now and we can refactor when the `panoptes-client/eras-client` works for this purpose.

I'm working on the `[panoptes-js](https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-panoptes-js)/erasAPI` at the moment, but that's the new client used predominantly by FEM, that's not a dependency in `mobile` at the moment.

So both of our commonly used clients, `panoptes-client` and `panoptes-js`, will soon have more ERAS functionality, but I think these one-off changes are good for now.

eatyourgreens replied:
[`panoptes-client`](https://www.npmjs.com/package/panoptes-client) is an old, stateful client with some odd bugs, because it's stateful. It keeps an undocumented cache of all downloaded resources, and prefers that cache to cut down on network requests. `import apiClient from panoptes-client/lib/api-client` imports an instance of the client with any previously requested resources stored in `apiClient._types` (I think, it's been a while since I looked at it.) This can lead to weird bugs where the client returns outdated copies of resources, instead of a fresh copy from the Panoptes API. There are undocumented  workarounds to bypass the internal cache, but it can be a pain if you're working with resources that change frequently (like users and user stats.)

[`@zooniverse/panoptes-js`](https://www.npmjs.com/package/@zooniverse/panoptes-js) is a newer client (about 5 years old), which is stateless, by design. It is responsible for requesting resources from Panoptes, but resource management is left to local state management in the parent app/component.

Both clients are old enough that they predate the `fetch` API in browsers and Node, and both use XHR, via superagent. With `fetch` now supported in all modern browsers, you're probably better off writing your own API code using `fetch`.

mcbouslog commented on src/api/index.js:
```diff
@@ -1,29 +1,26 @@
-import apiClient from 'panoptes-client/lib/api-client';
+import erasClient from 'panoptes-client/lib/eras-client';
```
Can be deleted since not using (doesn't support `project_contributions`) and using fetch instead.

mcbouslog reviewed (approved):
Changes LGTM, thank you!
